### ISSUE TITLE
Typesense Crash (json serialization in highlights)

### ISSUE THREAD

**Issue by Aldie** [typesense/typesense#190]
## Description

Typesense crashed under heavy load.   I wish I had more details for you - perhaps the backtrace can give you some insight.  This was running on BoardGameGeek with 2.8 million documents -   

## Metadata
```
E20210119 06:05:33.623625     1 typesense_server_utils.cpp:54] Typesense crashed...
E20210119 06:05:33.624555     1 typesense_server_utils.cpp:56] /opt/typesense-server(_Z11catch_crashi+0x1e) [0x82d51e]
E20210119 06:05:33.624577     1 typesense_server_utils.cpp:56] /lib/x86_64-linux-gnu/libc.so.6(+0x354b0) [0x7f22f97ef4b0]
E20210119 06:05:33.624585     1 typesense_server_utils.cpp:56] /opt/typesense-server(_ZN10Collection16highlight_resultERK5fieldRKSt6vectorIS3_IP8art_leafSaIS5_EESaIS7_EEPK2KVRKN8nlohmann10basic_jsonISt3mapS3
_NSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEblmdSaNSF_14adl_serializerES3_IhSaIhEEEER11StringUtilsmmbRKSN_SX_RNS_11highlight_tE+0x152) [0x76e182]
E20210119 06:05:33.624609     1 typesense_server_utils.cpp:56] /opt/typesense-server(_ZN10Collection6searchERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt6vectorIS5_SaIS5_EES7_SC_RKS8_I7sort_bySa
ISD_EEimm14token_orderingbmRKN3spp15sparse_hash_setIS5_NSJ_8spp_hashIS5_EESt8equal_toIS5_ENSJ_27libc_allocator_with_reallocIS5_EEEEST_mS7_mmS7_mS7_S7_SC_mS7_S7_S8_ImSaImEEm+0x421f) [0x77560f]
E20210119 06:05:33.624621     1 typesense_server_utils.cpp:56] /opt/typesense-server(_Z10get_searchR8http_reqR8http_res+0x2d44) [0x7c72b4]
E20210119 06:05:33.624631     1 typesense_server_utils.cpp:56] /opt/typesense-server(_ZN10HttpServer15process_requestEP8http_reqP8http_resP10route_pathPK24h2o_custom_req_handler_t+0x77) [0x7da877]
E20210119 06:05:33.624642     1 typesense_server_utils.cpp:56] /opt/typesense-server(_ZN10HttpServer17catch_all_handlerEP16st_h2o_handler_tP12st_h2o_req_t+0xc68) [0x7de5e8]
E20210119 06:05:33.624653     1 typesense_server_utils.cpp:56] /opt/typesense-server() [0xc03e84]
E20210119 06:05:33.624663     1 typesense_server_utils.cpp:56] /opt/typesense-server() [0xc0c353]
E20210119 06:05:33.624675     1 typesense_server_utils.cpp:56] /opt/typesense-server() [0xbf4e19]
E20210119 06:05:33.624686     1 typesense_server_utils.cpp:56] /opt/typesense-server() [0xbf4f58]
E20210119 06:05:33.624696     1 typesense_server_utils.cpp:56] /opt/typesense-server(h2o_evloop_run+0x2f) [0xbf724f]
E20210119 06:05:33.624703     1 typesense_server_utils.cpp:56] /opt/typesense-server(_ZN10HttpServer3runEP16ReplicationState+0x1b1) [0x7db1f1]
E20210119 06:05:33.624712     1 typesense_server_utils.cpp:56] /opt/typesense-server(_Z10run_serverRK6ConfigRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPFvvE+0xe12) [0x830c12]
E20210119 06:05:33.624720     1 typesense_server_utils.cpp:56] /opt/typesense-server(main+0x3f8) [0x83d268]
E20210119 06:05:33.624728     1 typesense_server_utils.cpp:56] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7f22f97da830]
E20210119 06:05:33.624737     1 typesense_server_utils.cpp:56] /opt/typesense-server() [0x72eaed]
```

**Typsense Version**: 0.18

**OS**: CentOS-7 Host / Running Docker image

**Comment by jasonbosco** [comment]
@Aldie Based on the stack trace, this looks like a JSON serialization issue, very similar to https://github.com/typesense/typesense/issues/148#issuecomment-727124310. Coincidentally, I initially assumed that issue was also because of high traffic, it later turned out to be because of a specific set of characters in the dataset I was using, that was throwing off JSON serialization. 

To debug this, we need to find out which set of characters in your dataset causes the server to crash. Here's one way to do this: write a script that takes the first 3 characters of the primary field searches are usually performed against, for each of the 2.8M records, and query the dataset with these 2.8M 3-character sequences, one at a time, to see if any of those crash the server. If you're able to isolate the particular search query that causes the crashes this way, that would help. 

Alternatively, if you're able to share this dataset with me (may be an S3 link via email - contact at typesense dot org), I can try debugging using it.

**Comment by Aldie** [comment]
@jasonbosco I tried reproducing the crash without success.  I did the 3 character search, I also tried a 4 character search using the titles we had.  I then queried the typesense server by document id and loaded every single document without a problem.  Any other ideas?

**Comment by jasonbosco** [comment]
@Aldie We've added more detailed stack traces in the upcoming version of Typesense, which will help us narrow down the cause of this issue faster. Would you be open to running an RC build of v0.19.0 in your stack for a bit and letting me know if you see the issue again?

If so, do you use the docker / deb / rpm package in production? I can get you the appropriate build.

**Comment by jasonbosco** [comment]
@Aldie We just published v0.19.0 that fixes a few issues that could trigger crashes. We've also added more debugging symbols to the stack trace. Could you upgrade to v0.19.0 and if/when the issue happens, share the detailed stack trace?

**Comment by kishorenc** [comment]
Going to close this for now, since I'm pretty sure this is fixed in v0.19.0 -- feel free to reopen if it occurs again. There is also much better stack traces now so that will certainly help.
